fuse/dl/models/backbones/backbone_transformer.py

```diff
@@ -36,12 +38,116 @@ def __init__(
-        :return: [batch_size, num_tokens + 1, token_dim] shaped tensor, where the first token is the CLS token
+        :return: [batch_size, num_tokens + num_cls_tokens, token_dim] shaped tensor, where the first tokens are the CLS tokens
         """
         b, n, _ = x.shape
-        cls_tokens = repeat(self.cls_token, "1 a d -> b a d", b=b)
+        cls_tokens = repeat(self.cls_tokens, "1 a d -> b a d", b=b)
         x = torch.cat((cls_tokens, x), dim=1)
         x += self.pos_embedding
         x = self.dropout(x)
         x = self.transformer(x)
         return x
+
+
+class CrossAttentionTransformerEncoder(nn.Module):
+    """
+    CrossAttentionTransformerEncoder backbone model based on x-transformers library.
+
+    Input:
+        two sequences 'seq_a, seq_b' with shapes [batch_size, len(seq_a)], [batch_size, len(seq_b)] respectively.
+    Output:
+        features tensor with shape [batch_size, output_dim]
+
+    see the following blog post for more info regarding cross attention in transformers:
+        https://vaclavkosar.com/ml/cross-attention-in-transformer-architecture
+    """
+
+    def __init__(
+        self,
+        emb_dim: int,
+        num_tokens_a: int,
+        num_tokens_b: int,
+        max_seq_len_a: int,
+        max_seq_len_b: int,
+        depth_a: int = 6,
+        depth_b: int = 6,
+        depth_cross_attn: int = 6,
+        heads_a: int = 9,
+        heads_b: int = 9,
+        output_dim: Optional[int] = None,
+        context: str = "seq_b",
+    ):
+        """
+        :param emb_dim: inner model dimension
+        :param num_tokens_a: number of tokens of the first sequence
+        :param num_tokens_b: number of tokens of the second sequence
+        :param max_seq_len_a: the maximum length of the first sequence
+        :param max_seq_len_b: the maximum length of the second sequence
+        :param depth_a: first sequence encoder's depth
+        :param depth_b: second sequence encoder's depth
+        :param depth_cross_attn: cross attender(s)' length
+        :param heads_a: number of attention heads for the first sequence's encoder
+        :param heads_b: number of attention heads for the second sequence's encoder
+        :param output_dim: (optional) model's output dimension. if not give the emb dim will be used as default.
+        :param context: which sequence will be used as context in the cross attention module:
+                        "seq_a": the first sequence will be used as a context
+                        "seq_b": the second sequence will be used as a context
+                        "both": will use two cross attention modules to take each one of the sequences as a context to the other one.
+        """
+        super().__init__()
+
+        if output_dim is None:
+            output_dim = emb_dim
+
+        assert context in ["seq_a", "seq_b", "both"]
+        self._context = context
+
+        # init sequences' encoders
+        self.enc_a = TransformerWrapper(
```

Comment:
Are there any additional arguments to TransformerWrapper or Encoder that it we should expose?

Reply:
Yes. actually there are a lot more.
I will add them as kwargs 😄